Process incoming frames before sending frames for Http2Connection.

danespinosa opened the pull request:
Moves the ProcessIncomingFramesAsync method before sending the preface.
This fixes a deadlock when using a NamedPipe as a transport where
incoming frames need to be read before being able to write.
Also rename _expectingSettingsAck to _receivedSettingsAck to clearly
specify that the settings frame has being received.

Fix #73097

Commit:
Process incoming frames before sending frames for Http2Connection.

Moves the ProcessIncomingFramesAsync method before sending the preface.
This fixes a deadlock when using a NamedPipe as a transport where
incoming frames need to be read before being able to write.
Also rename _expectingSettingsAck to _receivedSettingsAck to clearly
specify that the settings frame has being received.

Fix #73097

danespinosa commented:
Tests for this specific edge case would require me to implement an HTTP2 Server based on a NamedPipe transport. If necessary I think I could leverage the current infrastructure and extend it to be able to create a NamedPipe Server but I think what really matters is that none of the current tests should break. I believe that creating a NamedPipe server would be a bit cumbersome but doable after some iterations. Let me know if that would be needed please.

MihaZupan commented:
/azp run runtime-libraries-coreclr outerloop

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -222,11 +222,12 @@ public async ValueTask SetupAsync(CancellationToken cancellationToken)
                 BinaryPrimitives.WriteUInt32BigEndian(_outgoingBuffer.AvailableSpan, windowUpdateAmount);
                 _outgoingBuffer.Commit(4);
 
+                _receivedSettingsAck = false;
```
Nit: This is redundant now that the condition is inverted

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -222,11 +222,12 @@ public async ValueTask SetupAsync(CancellationToken cancellationToken)
                 _outgoingBuffer.Commit(4);
 
+                _receivedSettingsAck = false;
+                _ = ProcessIncomingFramesAsync();
```
Can you please add a comment here explaining that calling `ProcessIncoming` before `WriteAsync` is an intentional implementation detail to support named pipes?

danespinosa replied:
How does this sound?

```csharp
// Processing the incoming frames before sending the client preface and SETTINGS is necessary when using a NamedPipe as a transport.
// If the preface and SETTINGS coming from the server are not read first the below WriteAsync and the ProcessIncomingFramesAsync fall into a deadlock.
```

Also, there are 2 tests that I could clone and add a test with a NamedPipe stream.

ConnectCallback_ConnectionPrefix_Success clone and add **ConnectCallback_NamedPipeConnectionPrefix_Success**
ConnectCallback_UseUnixDomainSocket_Success clone and add **ConnectCallback_UseNamedPipeSocket_Success**

would that help? If the order of things ever change those tests could easily catch the issue.

MihaZupan replied:
That looks good to me.
The tests would be appreciated to prevent future regressions.

danespinosa replied:
Hi @MihaZupan  I have created a test that does detect the regresion, the only problem is that it does fail for HTTP1 since I didn't fix anything regarding HTTP1. So I guess the HTTP1 connection would need a similar fix. 
Should I work on that fix too or can the test target only http2 via a test attribute on the method?


<img width="845" alt="image" src="https://user-images.githubusercontent.com/30415120/182974960-58537eea-1fb8-4f4a-8df4-ead89e177851.png">

JamesNK replied:
IMO no one has asked for http/1.1 so leave it unchanged. Support can be added in the future when a need is identified.

danespinosa replied:
I modified the test so it would only be part of the http2 test flow only.
If a regression gets introduced the ConnectCallback_UseNamedPipe_Success test would fail.
I have tested that by reverting my change and the test does detect the regression.

danespinosa replied:
Just as an FYI I tested the new library on my test project and it works like a charm too :) Thanks a lot @MihaZupan  and @JamesNK  for your help

MihaZupan reviewed (approved):
LGTM assuming the existing tests pass.

antonfirsov commented:
Shouldn't we try to add a regression test based on the repro in #73097? Or it's not worth the trouble?

Commit:
Addressed PR Feedback.

Added test that will prevent regresions when using Http2 with NamedPipe as a transport.
Addressed PR Feedback

Fix #73307

Commit:
Update ConnectCallback_UseNamedPipe_Success test

Fix #73307

Commit:
Moved ConnectCallback_UseNamedPipe_Success test to
SocketsHttpHandlerTest_ConnectCallback_Http2 class

Fix #73307

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73307) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73307) before we can accept your contribution.<br/><br/>:x: danespinosa [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73307)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=73307) it.</sub>

JamesNK reviewed (approved):
LGTM
